src/cargo/core/compiler/mod.rs

```diff
@@ -1046,32 +1070,88 @@ impl Kind {
     }
 }
 
-fn assert_is_empty(line: &str) -> CargoResult<()> {
-    if !line.is_empty() {
-        Err(internal(&format!(
-            "compiler stdout is not empty: `{}`",
-            line
-        )))
-    } else {
-        Ok(())
-    }
+fn on_stdout_line(
+    state: &JobState<'_>,
+    line: &str,
+    _package_id: PackageId,
+    _target: &Target,
+) -> CargoResult<()> {
+    state.stdout(line);
+    Ok(())
 }
 
-fn json_stderr(line: &str, package_id: PackageId, target: &Target) -> CargoResult<()> {
-    // Stderr from rustc/rustdoc can have a mix of JSON and non-JSON output.
-    if line.starts_with('{') {
-        // Handle JSON lines.
-        let compiler_message = serde_json::from_str(line)
-            .map_err(|_| internal(&format!("compiler produced invalid json: `{}`", line)))?;
-
-        machine_message::emit(&machine_message::FromCompiler {
-            package_id,
-            target,
-            message: compiler_message,
-        });
-    } else {
-        // Forward non-JSON to stderr.
-        writeln!(io::stderr(), "{}", line)?;
+fn on_stderr_line(
+    state: &JobState<'_>,
+    line: &str,
+    package_id: PackageId,
+    target: &Target,
+    extract_rendered_errors: bool,
+    look_for_metadata_directive: bool,
+) -> CargoResult<()> {
+    // We primarily want to use this function to process JSON messages from
+    // rustc. The compiler should always print one JSON message per line, and
+    // otherwise it may have other output intermingled (think RUST_LOG or
+    // something like that), so skip over everything that doesn't look like a
+    // JSON message.
+    if !line.starts_with('{') {
```

Comment:
How confident are you that non-JSON output will never start with `{` (particularly with human error-format)? I haven't been able to think of a way to confuse it, but I'm concerned that this could inadvertently interpret a line as json.  The most likely case would be RUSTC_LOG debug output, though that might be a bit far fetched.

Reply:
I'm only semi-reasonably confident on this point. It's sort of an implicit contract we have currently, but it seems reasonable to gracefully handle lines that start with `{` yet aren't JSON and just print them out anyway instead of aborting compilation.